### PR TITLE
Elbv2-Deprecated SSL policies

### PR DESCRIPTION
**Pull request by alphadev4**

### CODE REVIEW

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            if (!describeLoadBalancers) return rcb();
+
+            if (describeLoadBalancers.err || !describeLoadBalancers.data) {
+                helpers.addResult(results, 3,
+                    `Unable to query for load balancers: ${helpers.addError(describeLoadBalancers)}`,region);
+                return rcb();
+            }
+
+            if (!describeLoadBalancers.data.length) {
+                helpers.addResult(results, 0, 'No load balancers found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:52]:
```suggestion
                helpers.addResult(results, 0, 'No Application or Network load balancers found', region);
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            for (var alb of describeLoadBalancers.data){
+                if (!alb.DNSName) continue;
+
+                var depPolicies = [];
+                var SslPolicy = false;
+                var describeListeners = helpers.addSource(cache, source,
+                    ['elbv2', 'describeListeners', region, alb.DNSName]);
+
+                if (describeListeners.err || !describeListeners.data || !describeListeners.data.Listeners) {
+                    helpers.addResult(results, 3, `Unable to query for Listeners: ${helpers.addError(describeListeners)}`, region);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:65]: resource is missing

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                    continue;
+                }
+
+                if (!describeListeners.data.Listeners.length) {
+                    helpers.addResult(results, 0, 'No Listeners found', region, alb.LoadBalancerArn);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:70]:
```suggestion
                    helpers.addResult(results, 0, 'No Listeners found for load balancer', region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                describeListeners.data.Listeners.forEach(function(listener){
+                    if (listener.SslPolicy) {
+                        SslPolicy = true;
+                        if (deprecatedALBPolicies.includes(listener.SslPolicy) || deprecatedNLBPolicies.includes(listener.SslPolicy)) {
+                            depPolicies.push(listener.SslPolicy);
+                        }
+                    }
+                });
+                if (!SslPolicy){
+                    helpers.addResult(results, 0,'No SSL policies found', region, alb.LoadBalancerArn);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:83]:
```suggestion
                    helpers.addResult(results, 0,'No SSL policies found load balancer', region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                } else if (depPolicies && depPolicies.length){
+                    helpers.addResult(results, 2, `The listeners on "${alb.LoadBalancerName}" are using following deprecated policies ` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:85]:
```suggestion
                    helpers.addResult(results, 2, `Load balancer listeners are using these deprecated policies :` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                } else if (depPolicies && !depPolicies.length){
+                    helpers.addResult(results, 0, `All listeners on "${alb.LoadBalancerName}" are using current SSL policies`, region, alb.LoadBalancerArn);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:87]:
```suggestion
                } else {
                    helpers.addResult(results, 0, 'Load balancer listeners are using current SSL policies', region, alb.LoadBalancerArn);
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+describe('elbv2DeprecatedSslPolicies', function () {
+    describe('run', function () {
+
+        it('should PASS if No Application Load Balancers found', function (done) {
+            const cache = createCache([]);
+            elbv2DeprecatedSslPolicies.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('No load balancers found');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js:132]:
```suggestion
                expect(results[0].message).to.include('No Application or Network load balancers found');
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                expect(results[0].status).to.equal(3);
+                done();
+            });
+        });
+
+        it('should PASS if no Listeners found', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[2]);
+            elbv2DeprecatedSslPolicies.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('No Listeners found');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js:162]:
```suggestion
                expect(results[0].message).to.include('No Listeners found for load balancer');
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                describeListeners.data.Listeners.forEach(function(listener){
+                    if (listener.SslPolicy) {
+                        SslPolicy = true;
+                        if (deprecatedALBPolicies.includes(listener.SslPolicy) || deprecatedNLBPolicies.includes(listener.SslPolicy)) {
+                            depPolicies.push(listener.SslPolicy);
+                        }
+                    }
+                });
+                if (!SslPolicy){
+                    helpers.addResult(results, 0,'No SSL policies found load balancer', region, alb.LoadBalancerArn);
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:83]:
```suggestion
                    helpers.addResult(results, 0,'No SSL policies found for load balancer', region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+                expect(results[0].status).to.equal(0);
+                done();
+            });
+        });
+
+        it('should PASS if no SSL policies found', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[1]);
+            elbv2DeprecatedSslPolicies.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('No SSL policies found');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js:172]:
```suggestion
                expect(results[0].message).to.include('No SSL policies found for load balancer');
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+                expect(results[0].status).to.equal(2);
+                done();
+            });
+        });
+
+        it('should PASS if all listeners on Elbv2 are using current policies', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[3]);
+            elbv2DeprecatedSslPolicies.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('are using current SSL policies');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.spec.js:192]:
```suggestion
                expect(results[0].message).to.include('listeners are using current SSL policies');
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+            for (var alb of describeLoadBalancers.data){
+                if (!alb.DNSName) continue;
+
+                var depPolicies = [];
+                var SslPolicy = false;
+                var describeListeners = helpers.addSource(cache, source,
+                    ['elbv2', 'describeListeners', region, alb.DNSName]);
+
+                if (describeListeners.err || !describeListeners.data || !describeListeners.data.Listeners) {
+                    helpers.addResult(results, 3, `Unable to query for Listeners: ${helpers.addError(describeListeners)}`, region);
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:65]:
```suggestion
                    helpers.addResult(results, 3, `Unable to query for Listeners: ${helpers.addError(describeListeners)}`, region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+                        SslPolicy = true;
+                        if (deprecatedALBPolicies.includes(listener.SslPolicy) || deprecatedNLBPolicies.includes(listener.SslPolicy)) {
+                            depPolicies.push(listener.SslPolicy);
+                        }
+                    }
+                });
+                if (!SslPolicy){
+                    helpers.addResult(results, 0,'No SSL policies found for load balancer', region, alb.LoadBalancerArn);
+                } else if (depPolicies && depPolicies.length){
+                    helpers.addResult(results, 2, `Load balancer listeners are using these deprecated policies :` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:85]:
```suggestion
                    helpers.addResult(results, 2, `Load balancer listeners are using these deprecated policies : ` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                        SslPolicy = true;
+                        if (deprecatedALBPolicies.includes(listener.SslPolicy) || deprecatedNLBPolicies.includes(listener.SslPolicy)) {
+                            depPolicies.push(listener.SslPolicy);
+                        }
+                    }
+                });
+                if (!SslPolicy){
+                    helpers.addResult(results, 0,'No SSL policies found for load balancer', region, alb.LoadBalancerArn);
+                } else if (depPolicies && depPolicies.length){
+                    helpers.addResult(results, 2, `Load balancer listeners are using these deprecated policies : ` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2DeprecatedSslPolicies.js:85]:
```suggestion
                    helpers.addResult(results, 2, 'Load balancer listeners are using these deprecated policies : ' + depPolicies.join(', '), region, alb.LoadBalancerArn);
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**